### ISSUE TITLE
as.integer not equivalent in PostgreSQL

### ISSUE THREAD

**Issue by iangow** [tidyverse/dbplyr#381]
**Brief description of the problem**

`as.integer()` produces different behaviour in PostgreSQL from what it produces in R and SQLite. In PostgreSQL it seems to round *up* for certain values (see `51` in reprex below).

I think if `CAST(x AS integer)` is desired, users should use SQL; `as.integer()` should translate to equivalent in SQL (in PostgreSQL, this would seem to require `floor()`).

``` r
library(dplyr, warn.conflicts = FALSE)
library(DBI)

conn_pg <- dbConnect(RPostgres::Postgres())
conn_lite <- dbConnect(RSQLite::SQLite())

df_lite <- 
  tbl(conn_lite, sql("SELECT 5080 AS sic")) %>%
  mutate(sic2 = as.integer(sic/100.0))

df_lite %>%
  show_query()
#> <SQL>
#> SELECT `sic`, CAST(`sic` / 100.0 AS INTEGER) AS `sic2`
#> FROM (SELECT 5080 AS sic)

df_lite
#> # Source:   lazy query [?? x 2]
#> # Database: sqlite 3.29.0 []
#>     sic  sic2
#>   <int> <int>
#> 1  5080    50

df_pg <- 
  tbl(conn_pg, sql("SELECT 5080 AS sic")) %>%
  mutate(sic2 = as.integer(sic/100.0))

df_pg %>%
  show_query()
#> <SQL>
#> SELECT "sic", CAST("sic" / 100.0 AS INTEGER) AS "sic2"
#> FROM (SELECT 5080 AS sic) "dbplyr_001"

df_pg
#> # Source:   lazy query [?? x 2]
#> # Database: postgres [igow@10.101.13.99:5432/crsp]
#>     sic  sic2
#>   <int> <int>
#> 1  5080    51

as.integer(5080.0/100)
#> [1] 50
```

<sup>Created on 2019-12-09 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by hadley** [comment]
I don't think `floor()` will work because `floor("a")` and `as.integer("a")` are rather different. I think this may just be a case where you have to expect differences between R and SQL, and choose your functions appropriately (and `floor()` or `trunc()` would more clearly express your intent here)

**Comment by iangow** [comment]
So it's not clear that `as.integer()` should exist in `dplyr` for PostgreSQL backends, as it implicitly promises something it doesn't deliver (i.e., a reasonable approximation of the.behaviour of the R function). Seems better to throw an error "`as.integer()` is not a PostgreSQL function" than to quietly do something unexpected.

**Comment by iangow** [comment]
That said, I mostly use `as.integer` to cast, say, `10.0` to `10L`. And here it's bevaviour works fine. It was using it with the example above that created the surprisingly hard-to-detect issue above.

**Comment by iangow** [comment]
> I don't think `floor()` will work because `floor("a")` and `as.integer("a")` are rather different. 

Indeed:

```
crsp=# SELECT floor('a');
ERROR:  invalid input syntax for type double precision: "a"
LINE 1: SELECT floor('a');

crsp=# SELECT CAST('a' AS INTEGER);
ERROR:  invalid input syntax for integer
```

I think *this* (i.e., `floor('a')`) may just be ~a~ **the better** case where "you have to expect differences between R and SQL, and choose your functions appropriately." Seems you would get an error either way.

**Comment by hadley** [comment]
`as.integer()` is primarily a coercion function. For your use case, you really should be using `floor()` or `trunc()`.

**Comment by iangow** [comment]
> `as.integer()` is primarily a coercion function. For your use case, you really should be using `floor()` or `trunc()`.

This is pretty much the only way I have used it. (The use above came from translating SAS code to R to promote usage of the latter). Even for coercion, I end up having to do checks along the lines of `all(as.integer(i) == i)` to make sure that nothing else is going on. (I think what *I* would like is a coerce-to-integer-if-nothing-changes-but-type-otherwise-return-an-error function and this is largely to put R data in the "correct" type before pushing PostgreSQL.)

Nonetheless I did some research using some test values. I think *even as a coercion function*, it's hard to get PostgreSQL or SQLite to behave like R. So this probably just needs to be added to a hypothetical documentation of "gotchas to be wary of when using `dplyr` against SQL sources".
